### PR TITLE
add conv2d conv1d forward function

### PR DESCRIPTION
**Pull request by FatJhon**
We have completed the forward  fuction of the conv1d, conv2d conv2d_depthwise operators.

### CODE REVIEW

**Comment by tongxin** [comment]
There are new conflicts needed to be resolve. Please see to that.

**Review by Bowen12992 — COMMENTED** [review]
Update the code to the newest to make all the cjecks work

File: src/flag_gems/ops/conv1d.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            input, (padding_width, padding_width), "constant", 0
+        )
+        padded_input = padded_input.contiguous()
+        _, _, padded_width_input = list(padded_input.shape)
+        out = torch.zeros(
+            n_out, c_out, width_out, dtype=input.dtype, device=input.device
+        )
+        grid = lambda meta: (
+            triton.cdiv(n_out, meta["BLOCK_N"]),
+            triton.cdiv(c_out, meta["BLOCK_O"]),
```
Comment (StrongSpoon) [inline src/flag_gems/ops/conv1d.py:120]: suggest using (n_out, c_out) as grid, since 3d pointer in triton has problems unsolved.

Comment (StrongSpoon) [inline src/flag_gems/ops/conv1d.py]: suggest using (n_out, c_out) as grid, since 3d pointer in triton has problems unsolved.

File: src/flag_gems/ops/conv1d.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    offset_cout = pid_cout * BLOCK_O + tl.arange(0, BLOCK_O)
+
+    accumulator = tl.zeros((BLOCK_N, BLOCK_O), dtype=tl.float32)
+    for i in range(0, width_out):
+        input_offset = (
+            (offset_n * c_input * padded_width_input)[:, None, None]
+            + (tl.arange(0, BLOCK_C_IN) * padded_width_input)[:, None]
+            + i * stride_width
+            + tl.arange(0, BLOCK_W)
+        )
```
Comment (StrongSpoon) [inline src/flag_gems/ops/conv1d.py]: doesn't support width_kernel that is not power of 2?

File: src/flag_gems/ops/conv1d.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        mask_weight = (offset_cout) < c_out and (tl.arange(0, BLOCK_C_IN))[
+            :, None, None
+        ] < c_input
+
+        input_value = tl.load(padded_input + input_offset, mask=mask_input, other=0)
+        weight_value = tl.load(weight + weight_offset, mask=mask_weight, other=0)
+        input_value = tl.reshape(input_value, (BLOCK_N, BLOCK_OUT_WEIGHT))
+        weight_value = tl.reshape(weight_value, (BLOCK_OUT_WEIGHT, BLOCK_O))
+        accumulator = tl.dot(input_value, weight_value, allow_tf32=False)
```
Comment (StrongSpoon) [inline src/flag_gems/ops/conv1d.py]: accumulator redefined

File: tests/test_reduction_ops.py [hunk 4]
```diff
@@ -689,3 +689,56 @@ def test_accuracy_index_select(shape, dim, dtype):
         res_out = torch.index_select(inp, dim, index)
 
     gems_assert_equal(res_out, ref_out)
+
+
+@pytest.mark.parametrize("shape", [(32, 2, 4)])
+@pytest.mark.parametrize("kernel", [(17, 2, 2)])
+@pytest.mark.parametrize("stride", [2])
+@pytest.mark.parametrize("padding", [1])
+@pytest.mark.parametrize("dtype", [torch.float32])
```
Comment (StrongSpoon) [inline tests/test_reduction_ops.py]: more test cases

**Review by StrongSpoon — COMMENTED** [review]
review ongoing

File: src/flag_gems/ops/conv2d.py [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+    input_cn_offset = (
+        offset_cn[:, None, None, None, None]
+        * c_input
+        * groups
+        * width_input
+        * height_input
+    )
+    input_ci_offset = offset_ci[:, None, None] * height_input * width_input
+    input_group_offset = (
+        pid_group[None, None, None, None] * c_input * width_input * height_input
```
Comment (StrongSpoon) [inline src/flag_gems/ops/conv2d.py]: initializing input_group_offset as a scalar is okay

File: src/flag_gems/ops/conv2d.py [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+        stride_height = ctx.stride_height
+
+        padding_height = ctx.padding_height
+
+        c_out = ctx.c_out
+        out_grad = out_grad.permute(0, 2, 3, 1).reshape(-1, c_out)
+        input_col = mm(out_grad, weight_reshape.T)
+
+        # return dx,None,None,None,None,None,None
+        conv2d_col2img[grid](
```
Comment (StrongSpoon) [inline src/flag_gems/ops/conv2d.py]: maybe we could fuse mm and col2img together as bwd kernel

Comment (StrongSpoon) [reply src/flag_gems/ops/conv2d.py]: but not necessarily

**Review by StrongSpoon — COMMENTED** [review]
please implement the performance tests.

File: src/flag_gems/ops/conv1d.py (hunk 2 above)
Comment (FatJhon) [reply src/flag_gems/ops/conv1d.py]: As Continuous data loading problem. Currently, block only supports powers of 2, so the h w of the kernel only supports this specification

File: src/flag_gems/ops/conv2d.py (hunk 6 above)
Comment (FatJhon) [reply src/flag_gems/ops/conv2d.py]: done

File: src/flag_gems/ops/__init__.py [hunk 7]
```diff
@@ -16,6 +16,9 @@
 from .bmm import bmm
 from .cat import cat
 from .clamp import clamp, clamp_tensor
+from .conv1d import conv1d
+from .conv2d import conv2d
+from .conv_depthwise2d import _conv_depthwise2d
```
Comment (StrongSpoon) [inline src/flag_gems/ops/__init__.py:21]: Is conv_depthwise2d implemented?

File: src/flag_gems/ops/conv1d.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    if isinstance(stride, (list, tuple)):
+        stride_width = stride[0]
+    else:
+        stride_width = stride
+
+    if isinstance(padding, (list, tuple)):
+        padding_width = padding[0]
+    else:
+        padding_width = padding
+    return conv2d(
```
Comment (StrongSpoon) [inline src/flag_gems/ops/conv1d.py:17]: I suggest implementing a kernel function for conv1d specifically. Calling conv2d will cost additional runtime.

File: src/flag_gems/ops/conv2d.py [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+    )[:, None]
+    weight_pointer += (
+        weight_n_stride * output_c_offset
+        + weight_n_stride * pid_group * out_per_group_c
+    )[None, :]
+
+    accum = tl.zeros((BLOCK_NI_HO_WO, BLOCK_CO), dtype=tl.float32)
+
+    for h in range(kernel_height):
+        for w in range(kernel_width):
```
Comment (StrongSpoon) [inline src/flag_gems/ops/conv2d.py]: since kernel_height and kernel_width are processed as loop iterator range, why not support those non-power-of-2?

File: src/flag_gems/ops/conv2d.py [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+        & (output_height_point_value < out_height)[:, None]
+        & (output_width_point_value < out_width)[:, None]
+    )
+
+    tl.store(output_pointer, accum, mask=output_mask)
+
+
+class Conv2d(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, input, weight, bias, stride, padding, dilation, groups):
```
Comment (iclementine) [inline src/flag_gems/ops/conv2d.py:381]: Dilation is not used at all, so it does not support dilation other than 1?

File: tests/test_reduction_ops.py (hunk 4 above)
Comment (FatJhon) [reply tests/test_reduction_ops.py]: added

File: src/flag_gems/ops/__init__.py (hunk 7 above)
Comment (FatJhon) [reply src/flag_gems/ops/__init__.py:21]: added

File: src/flag_gems/ops/conv2d.py (hunk 10 above)
Comment (FatJhon) [reply src/flag_gems/ops/conv2d.py:381]: Dilation is not used.

File: src/flag_gems/ops/conv2d.py [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+    Args:
+        in_size: Input size.
+        kernel_size: Kernel size.
+        stride: Stride.
+        padding: Padding.
+
+    Returns:
+        Output size of 2D convolution.
+    """
+    return (in_size + 2 * padding - kernel_size) // stride + 1
```
Comment (iclementine) [inline src/flag_gems/ops/conv2d.py]: Does it support asymmetric padding? I suppose not. Alright, torch's convolution does not support asymmetric padding.

File: src/flag_gems/ops/conv2d.py (hunk 10 above)
Comment (iclementine) [reply src/flag_gems/ops/conv2d.py:381]: I think we should raise an error if dilation > 1.

Comment (FatJhon) [reply src/flag_gems/ops/conv2d.py:381]: added dilation function

File: benchmark/core_shapes.yaml [hunk 12]
```diff
@@ -156,3 +156,17 @@ Benchmark:
     - [4096, 4096]
     - [64, 512, 512]
     - [1024, 1024, 1024]
+# default conv shape for input and weight stride padding groups
+# default Ni Ci Hi WI  Co Hk Wk stride padding groups
+ConvBenchmark:
+  shapes:
```
Comment (StrongSpoon) [inline benchmark/core_shapes.yaml:162]: 5 shapes are enough for core mode.

File: src/flag_gems/ops/conv2d.py [hunk 13]
```diff
@@ -0,0 +1,546 @@
...
+            out_c,
+            padding_height,
+            padding_width,
+            dilation_height,
+            dilation_width,
+        )
+
+        return (
+            input,
+            None,
```
Comment (iclementine) [inline src/flag_gems/ops/conv2d.py:535]: Gradient of weight and bias should also be computed.

Comment (iclementine) [reply src/flag_gems/ops/conv2d.py:535]: We leave it for future implementation.

**Review by iclementine — APPROVED** [review]

File: src/flag_gems/ops/conv2d.py [hunk 14]
```diff
@@ -0,0 +1,546 @@
...
+            )
+            input_mask = (
+                (no_point_value < in_n)[:, None]
+                & (input_c_offset < input_c)[None, :]
+                & (0 <= input_height_offset)[:, None]
+                & (input_height_offset < input_height)[:, None]
+                & (0 <= input_width_offset)[:, None]
+                & (input_width_offset < input_width)[:, None]
+            )
+            tl.atomic_add(curr_input_pointer, accum, input_mask)
```
Comment (iclementine) [inline src/flag_gems/ops/conv2d.py:376]: We can skip atomic_add by transforming conv_backward into a conv.
